gcc/rust/lint/rust-lint-marklive-base.h

```diff
@@ -0,0 +1,214 @@
+// Copyright (C) 2021 Free Software Foundation, Inc.
+
+// This file is part of GCC.
+
+// GCC is free software; you can redistribute it and/or modify it under
+// the terms of the GNU General Public License as published by the Free
+// Software Foundation; either version 3, or (at your option) any later
+// version.
+
+// GCC is distributed in the hope that it will be useful, but WITHOUT ANY
+// WARRANTY; without even the implied warranty of MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+// for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with GCC; see the file COPYING3.  If not see
+// <http://www.gnu.org/licenses/>.
+
+#ifndef RUST_HIR_LIVENESS_BASE
+#define RUST_HIR_LIVENESS_BASE
+
+#include "rust-diagnostics.h"
+#include "rust-lint-marklive.h"
+#include "rust-lint-marklive-base.h"
+#include "rust-hir-visitor.h"
+#include "rust-hir-map.h"
+
+namespace Rust {
+namespace Analysis {
+
+class MarkLiveBase : public HIR::HIRVisitor
+{
+public:
+  virtual ~MarkLiveBase () {}
+  virtual void visit (HIR::Token &) override {}
```

Comment:
Is it really wanted to override everything from the parent visitor (which already has everything defined as no-op) ? Wouldn't it be easier to read to only override methods where the behavior is not no-op?

Reply:
It sounds fine by me. I'll remove unused one.

Reply:
Hi @dkm , It's kind of awkward. I got a lot of warnings after removing the unused override methods. Maybe keep the compile msg clean will be better?

Reply:
What are the messages?

Reply:
when I comment
```
  // virtual void visit (HIR::Lifetime &) override {}
  // virtual void visit (HIR::LifetimeParam &) override {}
```
I got this output
```
In file included from ../../gccrs/gcc/rust/lint/rust-lint-marklive-base.h:25,
                 from ../../gccrs/gcc/rust/lint/rust-lint-marklive.h:25,
                 from ../../gccrs/gcc/rust/rust-session-manager.cc:52:
../../gccrs/gcc/rust/hir/tree/rust-hir-visitor.h:32:16: warning: 'virtual void Rust::HIR::HIRVisitor::visit(Rust::HIR::LifetimeParam&)' was hidden [-Woverloaded-virtual]
   32 |   virtual void visit (LifetimeParam &lifetime_param) = 0;
      |                ^~~~~
In file included from ../../gccrs/gcc/rust/lint/rust-lint-marklive.h:25,
                 from ../../gccrs/gcc/rust/rust-session-manager.cc:52:
../../gccrs/gcc/rust/lint/rust-lint-marklive-base.h:196:16: warning:   by 'Rust::Analysis::MarkLiveBase::visit' [-Woverloaded-virtual]
  196 |   virtual void visit (HIR::BareFunctionType &) override {}
      |                ^~~~~
In file included from ../../gccrs/gcc/rust/lint/rust-lint-marklive-base.h:25,
                 from ../../gccrs/gcc/rust/lint/rust-lint-marklive.h:25,
                 from ../../gccrs/gcc/rust/rust-session-manager.cc:52:
../../gccrs/gcc/rust/hir/tree/rust-hir-visitor.h:31:16: warning: 'virtual void Rust::HIR::HIRVisitor::visit(Rust::HIR::Lifetime&)' was hidden [-Woverloaded-virtual]
   31 |   virtual void visit (Lifetime &lifetime) = 0;
      |                ^~~~~
In file included from ../../gccrs/gcc/rust/lint/rust-lint-marklive.h:25,
                 from ../../gccrs/gcc/rust/rust-session-manager.cc:52:
../../gccrs/gcc/rust/lint/rust-lint-marklive-base.h:196:16: warning:   by 'Rust::Analysis::MarkLiveBase::visit' [-Woverloaded-virtual]
  196 |   virtual void visit (HIR::BareFunctionType &) override {}
      |                ^~~~~
In file included from ../../gccrs/gcc/rust/rust-session-manager.cc:53:
../../gccrs/gcc/rust/lint/rust-lint-scan-deadcode.h: In static member function 'static void Rust::Analysis::ScanDeadcode::Scan(Rust::HIR::Crate&, std::set<unsigned int>)':
../../gccrs/gcc/rust/lint/rust-lint-scan-deadcode.h:39:18: error: cannot declare variable 'sdc' to be of abstract type 'Rust::Analysis::ScanDeadcode'
   39 |     ScanDeadcode sdc (live_symbols);
      |                  ^~~
../../gccrs/gcc/rust/lint/rust-lint-scan-deadcode.h:32:7: note:   because the following virtual functions are pure within 'Rust::Analysis::ScanDeadcode':
   32 | class ScanDeadcode : public MarkLiveBase
      |       ^~~~~~~~~~~~
In file included from ../../gccrs/gcc/rust/lint/rust-lint-marklive-base.h:25,
                 from ../../gccrs/gcc/rust/lint/rust-lint-marklive.h:25,
                 from ../../gccrs/gcc/rust/rust-session-manager.cc:52:
../../gccrs/gcc/rust/hir/tree/rust-hir-visitor.h:31:16: note: 	'virtual void Rust::HIR::HIRVisitor::visit(Rust::HIR::Lifetime&)'
   31 |   virtual void visit (Lifetime &lifetime) = 0;
      |                ^~~~~
../../gccrs/gcc/rust/hir/tree/rust-hir-visitor.h:32:16: note: 	'virtual void Rust::HIR::HIRVisitor::visit(Rust::HIR::LifetimeParam&)'
   32 |   virtual void visit (LifetimeParam &lifetime_param) = 0;
      |                ^~~~~

```

One error occurs too, deleting the no-op will make the derived class keep abstract.

Reply:
I think you can use this to fix that warning: https://github.com/Rust-GCC/gccrs/blob/59f09fd5a48046ceba064bde2f230386fbc3700f/gcc/rust/resolve/rust-ast-resolve-expr.h#L33

Reply:
Actually, I did using the visit from base: 
https://github.com/Rust-GCC/gccrs/blob/59f09fd5a48046ceba064bde2f230386fbc3700f/gcc/rust/lint/rust-lint-scan-deadcode.h#L34

Reply:
In the case of method being pure virtual, then you need to have an override yes. Wondering why LifetimeParam is defined as pure virtual in HIRVisitor. I guess it could simply be made no-op by default?

Reply:
I'm not sure about that. All of the visit in HIRVisitor are pure virtual method.

Reply:
Ho, sorry, didn't see that. Then you need to add no-op override. @philberty any reason for having all methods pure-virtual ?